src/Http/Routing/src/Builder/EndpointRoutingApplicationBuilderExtensions.cs

```diff
@@ -95,5 +113,31 @@ private static void VerifyRoutingIsRegistered(IApplicationBuilder app)
                     "ConfigureServices(...)"));
             }
         }
+
+        private static void VerifyEndpointRoutingMiddlewareIsRegistered(IApplicationBuilder app, out DefaultEndpointRouteBuilder endpointRouteBuilder)
+        {
+            if (!app.Properties.TryGetValue(EndpointRouteBuilder, out var obj))
+            {
+                var message =
+                    $"{nameof(EndpointRoutingMiddleware)} must be added to the request execution pipeline before {nameof(EndpointMiddleware)}. " +
+                    $"Please add {nameof(EndpointRoutingMiddleware)} by calling '{nameof(IApplicationBuilder)}.{nameof(UseRouting)}' inside the call " +
+                    $"to 'Configure(...)' in the application startup code.";
+                throw new InvalidOperationException(message);
```

Comment:
Throwing if UseRouting is not present when UseEndpoints is called is nice.

What do you think of expanding this message to explain why routing should be first? I think it is useful to teach the dev about what is going on at each step. On the other hand the message is quite long already.

Reply:
What about the scenario when some middleware has set an endpoint without routing, and we just want to execute the set endpoint? This is what EndpointMiddleware use to do.

Reply:
> What do you think of expanding this message to explain why routing should be first? I think it is useful to teach the dev about what is going on at each step. On the other hand the message is quite long already.

I really don't personally think we can use exception messages to teach people things, other than that the current thing is wrong. If you have a suggestion of how to reword this or some specific language you want me to add, I'd be happy to try anyway. 


> What about the scenario when some middleware has set an endpoint without routing, and we just want to execute the set endpoint? This is what EndpointMiddleware use to do.

I don't want this to work. I don't want the endpoint middleware to be optional, or to show conflicting examples.

Reply:
> If you have a suggestion of how to reword this or some specific language you want me to add, I'd be happy to try anyway.

`$"{nameof(EndpointRoutingMiddleware)} matches routes setup by {nameof(EndpointMiddleware)} and so must be added to the request execution pipeline before {nameof(EndpointMiddleware)}."`

Not super happy with it but it explains a very little why we ask devs to do the things they must do.

Reply:
> I don't want the endpoint middleware to be optional, or to show conflicting examples.

I get that, but what about if someone has middleware that sets an endpoint. There is no longer a `UseEndpoint` (implicit or explicit) to execute it. This change feels like endpoints are being constrained to always being used with routing.

Reply:
> But what about if someone has middleware that sets an endpoint. There is no longer a UseEndpoint (implicit or explicit) to execute it.

Do we have a scenario for this? I think making the endpoint middleware both implicit and explicit is a decision we will be explaining forever. Making it explicit is a response to feedback that people don't understand how the pipeline works when its implicit.

I'm happy to discuss whatever options you think are improvements. One idea I have would be to add an overload of `UseEndpoints()` that doesn't register any endpoints and thus doesn't require `UseRouter`. 

This all seems so marginal to me since this is trivial to write on your own. I can't think of a reason why we would document or recommend that as a pattern.

Reply:
I kind of did something like that in this auth sample - https://github.com/aspnet/AspNetCore/pull/4279 - although the reason for setting the endpoints was metadata for the AuthZ middleware. The static file middleware would take care of executing the endpoints.

Reply:
> The static file middleware would take care of executing the endpoint.

Right, so the point is that someone executes the endpoint and doesn't call next - then you never reach this condition.

Reply:
Yes, in this example.

Basically my scenario is:
1. Middleware decides to send a response to the user and sets an endpoint rather than be terminal (maybe it wants more middleware to run after it)
2. Request gets to the end of the pipeline and the host throws an error saying their is an endpoint so the user must configure UseEndpoints
3. User adds UseEndpoints
4. Now ASP.NET Core throws an error saying the user must configure UseRouting, which user doesn't need/want

Reply:
Resolved this but updating the error message and adding a `UseEndpointExecutor()` method.